File: client/src/main/java/io/xol/chunkstories/gui/layer/config/LogPolicyAsk.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.gui.layer.config;

import io.xol.chunkstories.api.gui.Font;
import io.xol.chunkstories.api.gui.Gui;
import io.xol.chunkstories.api.gui.GuiDrawer;
import io.xol.chunkstories.api.util.Configuration;
import org.joml.Vector4f;

import io.xol.chunkstories.api.gui.Layer;
import io.xol.chunkstories.api.gui.elements.Button;

/** Asks the user if he wishes to have his logs uploaded to the game servers for debugging purposes */
//TODO anonymize those (strip C:\Users\... and such)
public class LogPolicyAsk extends Layer {
	public static final String logPolicyConfigNode = "client.game.logPolicy";

	private Configuration.OptionString option = gui.getClient().getConfiguration().get(logPolicyConfigNode);

	private Button acceptButton = new Button(this, 0, 0, 150, "#{logpolicy.accept}");
	private Button refuseButton = new Button(this, 0, 0, 150, "#{logpolicy.deny}");

	private String logPolicyExplanationText = gui.localization().getLocalizedString("logpolicy.asktext");

	public LogPolicyAsk(Gui gui, Layer parent) {
		super(gui, parent);

		this.acceptButton.setAction(new Runnable() {

			@Override
			public void run() {
				option.trySetting("send");
				gui.getClient().getConfiguration().save();
				LogPolicyAsk.this.gui.setTopLayer(parentLayer);
			}

		});

		this.refuseButton.setAction(new Runnable() {

			@Override
			public void run() {
				option.trySetting("dont");
				gui.getClient().getConfiguration().save();
				LogPolicyAsk.this.gui.setTopLayer(parentLayer);
			}

		});

		elements.add(acceptButton);
		elements.add(refuseButton);
	}

	@Override
	public void render(GuiDrawer drawer) {
		parentLayer.render(drawer);

		drawer.drawBox(0, 0, gui.getViewportWidth(), gui.getViewportHeight(), 0, 0, 0, 0, null, new Vector4f(0.0f, 0.0f, 0.0f, 0.5f));

		drawer.drawStringWithShadow(
				drawer.getFonts().getFont("LiberationSans-Regular__aa", 16 * 1), 30,
				gui.getViewportHeight() - 64,
				gui.getClient().getContent().localization().getLocalizedString("logpolicy.title"), -1,
				new Vector4f(1.0F));

		Font logPolicyTextFont = drawer.getFonts().getFont("LiberationSans-Regular__aa", 12);

		drawer.drawString(logPolicyTextFont, 30, gui.getViewportHeight() - 128, logPolicyExplanationText, width - 60, new Vector4f(1.0F));

		int buttonsSpacing = 4;
		int buttonsPlusSpacingLength = acceptButton.getWidth() + refuseButton.getWidth() + buttonsSpacing;

		acceptButton.setPosition(gui.getViewportWidth() / 2 - buttonsPlusSpacingLength / 2,
				gui.getViewportHeight() / 4 - 32);
		acceptButton.render(drawer);

		refuseButton.setPosition(
				gui.getViewportWidth() / 2 - buttonsPlusSpacingLength / 2 + buttonsSpacing + acceptButton.getWidth(),
				gui.getViewportHeight() / 4 - 32);
		refuseButton.render(drawer);
	}
}

File: client/src/main/java/xyz/chunkstories/gui/layer/config/LogPolicyAsk.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.gui.layer.config

import xyz.chunkstories.api.gui.Font
import xyz.chunkstories.api.gui.Gui
import xyz.chunkstories.api.gui.GuiDrawer
import xyz.chunkstories.api.util.configuration.Configuration
import org.joml.Vector4f

import xyz.chunkstories.api.gui.Layer
import xyz.chunkstories.api.gui.elements.Button

/** Asks the user if he wishes to have his logs uploaded to the game servers for debugging purposes  */
//TODO anonymize those (strip C:\Users\... and such)
class LogPolicyAsk(gui: Gui, parent: Layer) : Layer(gui, parent) {

    private val option = gui.client.configuration.get<Configuration.OptionString>(logPolicyConfigNode)

    private val acceptButton = Button(this, 0, 0, 150, "#{logpolicy.accept}")
    private val refuseButton = Button(this, 0, 0, 150, "#{logpolicy.deny}")

    private val logPolicyExplanationText = gui.localization().getLocalizedString("logpolicy.asktext")

    init {

        this.acceptButton.action = Runnable {
            option!!.trySetting("send")
            gui.client.configuration.save()
            this@LogPolicyAsk.gui.topLayer = parentLayer
        }

        this.refuseButton.action = Runnable {
            option!!.trySetting("dont")
            gui.client.configuration.save()
            this@LogPolicyAsk.gui.topLayer = parentLayer
        }

        elements.add(acceptButton)
        elements.add(refuseButton)
    }

    override fun render(drawer: GuiDrawer) {
        parentLayer?.render(drawer)

        drawer.drawBox(0, 0, gui.viewportWidth, gui.viewportHeight, 0f, 0f, 0f, 0f, null, Vector4f(0.0f, 0.0f, 0.0f, 0.5f))

        drawer.drawStringWithShadow(
                drawer.fonts.getFont("LiberationSans-Regular__aa", (16 * 1).toFloat()), 30,
                gui.viewportHeight - 64,
                gui.client.content.localization().getLocalizedString("logpolicy.title"), -1,
                Vector4f(1.0f))

        val logPolicyTextFont = drawer.fonts.getFont("LiberationSans-Regular__aa", 12f)

        drawer.drawString(logPolicyTextFont, 30, gui.viewportHeight - 128, logPolicyExplanationText, width - 60, Vector4f(1.0f))

        val buttonsSpacing = 4
        val buttonsPlusSpacingLength = acceptButton.width + refuseButton.width + buttonsSpacing

        acceptButton.setPosition(gui.viewportWidth / 2 - buttonsPlusSpacingLength / 2,
                gui.viewportHeight / 4 - 32)
        acceptButton.render(drawer)

        refuseButton.setPosition(
                gui.viewportWidth / 2 - buttonsPlusSpacingLength / 2 + buttonsSpacing + acceptButton.width,
                gui.viewportHeight / 4 - 32)
        refuseButton.render(drawer)
    }

    companion object {
        val logPolicyConfigNode = "client.game.logPolicy"
    }
}
